### DISCUSSION TITLE
Using Bootstrap with react-page

### DISCUSSION THREAD

**Discussion by ennovative-pl** [react-page/react-page#1259] [Q&A]
Hi, I am evaluating react-page for use in our existing production system. I like way that positioning is being done, ability to add custom plugins, next.js compability, being React and TS based, and general design ideas. 

Our use case is that we would be creating a lot of custom plugins of form elements (based on formik) and using react-page as a WYSIWYG form editor for formik-based forms. 

But the problem is, that our project is already quite large, and uses ... Bootstrap for positioning and controls. We would not be switching to mui at the moment for sure. 

As we need seamless integration and visual parity between different parts of our system, I am contemplating creating a fork and replacing mui with Bootstrap. What do you guys think about it? Is it even feasible, or positioning and other stuff iin react-page is so dependant on mui, that it is a task that is doomed to fail from the very beginning?

**Comment by SebastianStehle** [comment]
Why would you replace one dependency with another? 

Personally I am not a fan of the dependency to MUI and the general look. I would prefer not to have a dependency and a more cleaner / flat look. But the appearance is just a personal preference.

I think that today mui is mainly used in two places:

1. Icons
2. Uniform dialogs: https://github.com/vazco/uniforms

So I would:

1. Reduce the dependency to MUI in the toolbar on the right and rebuild these buttons with just svg elements and normal buttons. Choose good class names so that you can override them.
2. Use just SVGs for all icons to get rid of the dependency to MUI icons.
3. Make the theming system from uniforms available in the props of the react-editor.
4. Rebuilt all custom forms with uniforms: https://uniforms.tools/docs/examples-custom-fields. I am not sure if this is possible.

With this approach you reduce the dependency to MUI and you can easily use ant design or bootstrap or whatever.

**Comment by macrozone** [comment]
you don't need to to switch the rest of the app to mui, alltough you might want to tweak the themeing a bit, so it looks similar to the rest of your app. The additional bundle size is usually neglectable in the type of apps where react-page is used. Also be aware that react-page does not load mui and most of its code when used to just diplay content (readyOnly)

the visual appearance of mui is debateable, but it's one of the best component library and unlike bootstrap or antd, which are css based it does not leak into the rest of your application.

making react-page more agnostic to the ui would be nice, but I think many underestimate how challenging it is to do this properly and overestimate the value of this.

> Reduce the dependency to MUI in the toolbar on the right and rebuild these buttons with just svg elements and normal buttons. Choose good class names so that you can override them.

the right toolbar could be indeed be made without mui, but there are many other areas where it is used

> Use just SVGs for all icons to get rid of the dependency to MUI icons.

why? mui icon set is not bound to mui and supports tree-shaking, there is not much benefit of just inlining those svgs. 

> Make the theming system from uniforms available in the props of the react-editor.

the theming system of mui is available as props. uniforms has no real theming system, it has just some packages of common ui libraries, but those packages are copies from components and have slight differences. 

> Rebuilt all custom forms with uniforms: https://uniforms.tools/docs/examples-custom-fields. I am not sure if this is possible.

uniform is used already. There are however some (legacy) built-in CellPlugins like image and background that have a custom form built with mui. Its actually easy to replace those with custom CellPlugins that use an autoform. Its also possible to replace the content of an autoform with a custom element, so you could use the antd-version in theory.

**Reply by ennovative-pl** [reply]
My use case is quite specific, as the rest of the app is using forms in multiple places, and - the most important thing - allows heavy customization with themes. Bootstrap ones. Each user can choose a theme for his account, there is a possibility to even upload own theme into the application. It is because brand customization is really important part of my application. Two different users could have totally different look & feel. I want my form editing part to also use those themes. I have no choice - but to make everything use bootstrap classes so that colors, fonts and general feel would be uniform within my application from the user perspective. 

It would be great, if react-page would be little less opinionated when it comes to it's UI... I would gladly help with unwinding it a little bit, maybe create an abstraction layer for UI, so that developer could choose what kind of a library, or just even plain HTML elements he would like to use. A good example is react-awesome-query-builder, it has Bootstrap, Ant, MUI and vanilla versions, and it's quite easy to modify just UI layer... I am using this in my project also, and a year ago, when there was no Bootstrap version available yet, I was able to generally implement it in under a day... It seems, that in react-page this would be a "little" more time... 

Maybe we would be able to contribute in this area, if there would be consensus that it would be an useful addition to this project...

**Reply by ylnsnv** [reply]
Any news regarding this?
@ennovative-pl It would be a huge benefit to this amazing library...
Oh and btw, for me, I'd like to use react-page with tailwind customization.

**Reply by kisg** [reply]
@ylnsnv @ennovative-pl 

I opened a similar question a couple weeks ago, but for React Native (tamagui.dev to be exact): https://github.com/react-page/react-page/discussions/1297

We also would be interested in contributing, if there is a consensus on the implementation strategy.
